### ISSUE TITLE
checkpoint info access

### ISSUE THREAD

**Issue by ramagottfried** [rncm-prism/prism-samplernn#15]
hello! thank you for your work on this! I've been slowly making some progress with a data set I'm trying to train on. Still some ways to go, but still hopeful for an interesting results.

In a few of my tests, the accuracy level dove down very low after some time, but after tweaking hyper-parameters, the results are getting better.

However, some of the earlier incremental generation samples seem to be better than later (but of course still hoping it will improve!) -- is there an easy way to see what the accuracy and loss values were for those checkpoints?

I have verbose mode on since it's interesting follow the progress, but then the values from previous epochs are not accessible on the screen any more.  

I tried viewing some of the checkpoint and validation files in a text editor (vs code) but it seems to be a binary, or non utf-8 encoding.

thanks for any advice!

**Comment by ramagottfried** [comment]
related general practice question, if say epoch N has a pretty interesting generation output, and then the quality seems to drop off after a while, does it seem like a good idea to quit and resuming training from epoch N in hopes that it finds a better trajectory? 

(I'm using Adam optimizer at the moment but tempted to try the others to see if they work differently -- sample set size of 2640 files @ 8s each, current learning rate 0.001)

**Comment by relativeflux** [comment]
Hi, and thanks for your interest in the project. Indeed you're right, checkpoints are binary files. They're standard TensorFlow checkpoints (more precisely part of TensorFlow's Keras implementation), which can be loaded and inspected in TensorFlow itself, but the best way to investigate the training history is using TensorBoard. TensorBoard is your friend, honestly I cannot recommend it enough. To run it simply execute:

`tensorboard --logdir path/to/some/logdir`

Where `--logdir` points to the folder for your training session (we're assuming here that you are running it in a fresh console from the `prism-samplernn` top-level directory, and that your training logdir will be in the default location of `prism-samplernn/logdir/<train_id>`). So, if you have a training session with the id `foo` then you would point `--logdir` to `logdir/foo`. Each training run stores checkpoints in a timestamped subdirectory of that directory, but in addition to checkpoints it will also contain `train` and `validation` subdirectories, which store summaries of the training metrics (loss, accuracy), for display in TensorBoard (see screenshot below).

The TensorBoard UI is updated every 30 seconds by default (you can modify this value from inside TensorBoard itself).

For monitoring training I use a combination of TensorBoard, test audio generated during training and also a log directory storing the (non-verbose) training output. I generate the log as output from `nohup`, since I work on a remote machine (you could also use a tool like `screen`):

nohup python train.py \
  --id Beethoven_32_Piano_Sonatas-8s \
  --data_dir ~/datasets/Beethoven_32_Piano_Sonatas-8s \
  --config_file lvb-sonatas.config.json \
  --num_epochs 250 \
  --batch_size 128 \
  --verbose False \
  --resume False \
  --checkpoint_every 5 \
  --max_checkpoints 20 \
  --monitor val_loss \
  --early_stopping_patience 4 \
  > Beethoven_32_Piano_Sonatas-8s.log 2>&1 </dev/null &

TensorBoard has lots of features, but the one I mainly use is the Scalar page, which contains graphs showing the change in loss and accuracy over time. The most important one of these is the `epoch_loss` graph, which charts the progress of the training and validation losses over the epochs. Looking at the relationship between the training and validation losses is the best way to assess the progress of training. If the training loss continues to descend while the val loss either flatlines or starts to ascend then you're almost certainly overfitting - meaning the network is learning the target (including the noise in the target) too well, and can't generalise. This is usually because the network is too strong (too many neurons, too many layers), so you can either lower the RNN dimensionality, or use fewer RNN layers. Alternatively you can increase the size of your dataset, although this is often not possible (chunking with an overlap can help increase the number of chunks). The opposite situation, underfitting, is usually indicated by both the training and val losses flatlining or ascending. This means the network is not powerful enough, and so the solution is to add more layers and/or increase the dimensionality (and/or reduce the size of the dataset). Flatling can also occur if the loss gets stuck in a local minima, for which a recommended solution is to kill the training and resume with a lower learning rate (say, down from 0.001 to 0.0001), and/or a lower batch size (say, from 128 to 64).

Your dataset size seems about right, and I rarely start with a learning rate other than the default. I am currently training on the Beethoven 32 piano sonatas, in an attempt to repeat the original SampleRNN experiment. I will publish the results, but here's some short test output files, about 60 epochs in, temperature 0.95 (had to convert to .mp4 format for upload):

[lvb_32_pno_sonatas_temp_0.95_0.mp4](https://user-images.githubusercontent.com/17030640/113487529-903cf100-94b0-11eb-9357-ae93df3b44cf.mp4)
[lvb_32_pno_sonatas_temp_0.95_3.mp4](https://user-images.githubusercontent.com/17030640/113487569-ca0df780-94b0-11eb-8d13-20ea62dd6768.mp4)

What's nice is that you can hear some structure, there are phrases, and cadences (sort of). There's definitely some LVB in there, although it also sounds quite Schummanesque. to my ear!

**Comment by relativeflux** [comment]
@ramagottfried My guess is if you're seeing worse results later in the training, after getting some good output, the network has started to overfit - it will always have a tendency to do this, if you keep it training for too long after it's found a good fit. TensorBoard grpahs are invaluable for finding these inflection points in the training, and for selecting potentially good epochs.

**Comment by relativeflux** [comment]
Output from same epoch with slightly higher temperature:

[lvb_32_pno_sonatas_temp_0.975_1.mp4](https://user-images.githubusercontent.com/17030640/113491202-ee74ce80-94c6-11eb-87f1-40d4249f1b8f.mp4)

**Comment by ramagottfried** [comment]
@relativeflux great! thank you so much for your help! I finally got the tensorboard working, I'm using docker so just had to figure out the ports, and now looking at the graphs it does seem like the last training run was actually still improving when I killed it, even if the generated sounds were less interesting (more silence, and distorted sounds), hmm, well I started it again from my favorite checkpoint with 0.0001 learning rate (i'm using maybe a very small batch size of 16), seems to be doing well, let's see what happens...  and thanks for sending the piano examples! very nice to have examples for the repository I think -- and they really sounds great! I like all of the strange phrasing turns, almost like someone practicing, and then the ringing strings have a strange extra spectral resonance, super cool
